invoices/update.go

```diff
@@ -206,15 +229,91 @@ func updateMpp(ctx *invoiceUpdateCtx,
 		return &update, ctx.acceptRes(resultAccepted), nil
 	}
 
+	var (
+		htlcPreimages map[channeldb.CircuitKey]lntypes.Preimage
+		htlcPreimage  lntypes.Preimage
+	)
+	if ctx.amp != nil {
+		// Create a slice containing all the child descriptors to be
+		// used for reconstruction. This should include all HTLCs
+		// currently in the HTLC set, plus the incoming HTLC.
+		childDescs := make([]amp.ChildDesc, 0, 1+len(htlcSet))
+
+		// Add the new HTLC's child descriptor at index 0.
+		childDescs = append(childDescs, amp.ChildDesc{
+			Share: ctx.amp.RootShare(),
+			Index: ctx.amp.ChildIndex(),
+		})
+
+		// Next, construct an index mapping the position in childDescs
+		// to a circuit key for all preexisting HTLCs.
+		indexToCircuitKey := make(map[int]channeldb.CircuitKey)
+
+		// Add the child descriptor for each HTLC in the HTLC set,
+		// recording it's position within the slice.
+		var htlcSetIndex int
+		for circuitKey, htlc := range htlcSet {
+			childDescs = append(childDescs, amp.ChildDesc{
+				Share: htlc.AMP.Record.RootShare(),
+				Index: htlc.AMP.Record.ChildIndex(),
+			})
+			indexToCircuitKey[htlcSetIndex] = circuitKey
+			htlcSetIndex++
+		}
+
+		// Using the child descriptors, reconstruct the root seed and
+		// derive the child hash/preimage pairs for each of the HTLCs.
+		children := amp.ReconstructChildren(childDescs...)
+
+		// Construct a closure that will failure the existing HTLC set
+		// and the new HTLC in case reconstruction failed.
+		failReconstruction := func() *HtlcFailResolution {
+			update.State = &channeldb.InvoiceStateUpdateDesc{
+				NewState: channeldb.ContractCanceled,
+				SetID:    setID,
+			}
+			return ctx.failRes(ResultAmpReconstruction)
+		}
+
+		// Validate that the derived child preimages match the hash of
+		// each HTLC's respective hash.
+		if ctx.hash != children[0].Hash {
```

Comment:
Ah because this final HTLC won't yet have been committed during acceptance as HTLC part of the HTLC set of the main invoice?

Reply:
correct, when the final HTLC comes in we have the other `n-1` recorded and need to examine if this "completes the set" or not